mpi: Cleanup definitions in preperations of the lowering

AntonLydike opened the pull request:
This PR does not contains any tests, as they will be included in the PR that contains the lowering.

georgebisbas commented on xdsl/dialects/mpi.py:
```diff
@@ -311,15 +318,37 @@ def get(cls, status_obj: Operand, field: StatusTypeField):
 
 
+@irdl_op_definition
+class CommRank(MPIBaseOp):
```
Even small, simple tests

superlopuh commented on tests/dialects/test_mpi.py:
```diff
@@ -13,46 +14,70 @@
-  %8 = "mpi.status.get"(%6) {"field" = "MPI_SOURCE"} : (#mpi.status) -> si32
-  %9 = "mpi.status.get"(%6) {"field" = "MPI_ERROR"} : (#mpi.status) -> si32
-  %10 = "mpi.wait"(%4) : (#mpi.request) -> si32
+  "func.func"() ({
```
This string is getting to be quite long, and feels more and more like an integration test as opposed to a unit test. I'm not sure what is being tested here

webmiche replied:
Strong +1

webmiche commented:
I feel I once again need to comment that one big test is IMO not a great way to test as it does not give good debug info, if this test fails.

AntonLydike commented:
I guess this is the point where I admit that I don't really understand our testing strategy. 

If we have a test like this:

```python3
def test_mpi_ISend():
    alloc0 = Alloc.get(f64, 32, [100, 14, 14])
    dest = Constant.from_int_and_width(1, t_int)
    send = ISend.get(alloc0, dest, 1)

    assert send.operands[0] is alloc0.results[0]
    assert send.operands[1] is dest.results[0]
```

What value does this test provide? This test tests that the `.get` method passes its arguments to the correct place. 

Or to rephrase that question, what does it tell us when this singular test fails? It tells us that we changed the `.get` method of the `ISend` operation. But if we did that, we already *know* that we changed that method, because it's the only thing we could have done. For me, these tests don't cover any *interesting behavior*.

Contrast that with a test like this:

```
@irdl_op_definition
class ResultOp(Operation):
    name: str = "test.result_op"

    res: Annotated[OpResult, StringAttr]


def test_result_builder():
    op = ResultOp.build(result_types=[StringAttr("")])
    op.verify()
    assert op.res is op.results[0]
    assert [res.typ for res in op.results] == [StringAttr("")]
```

When specifically this test fails (and no other), we know that we screwed up something in how irdl handles results. That's valuable information! We can work with that!

I'd really appreciate some input on this.

webmiche commented:
So I guess the real answer is that we don't really have a testing strategy, because all of us never worked in such a big project and we only know somewhat what others do, not exactly why and how.

The reason for the tests you cite above is mainly to get lines covered and actually executed in order to catch bugs. And since we mainly have the get methods in the operations, we mainly test those.

I am not quite sure what you mean about the whole argument that this test doesn't tell us anything. It does tell us that the get method (the main interface of this operation) is working/breaking. This is quite crucial information, IMO, and it has helped us catch bugs and still does. 

Do you have a suggestion what we could change to get this better?

Also, you can extend your argument to the tests you have in this PR, IMO, so do they not tell us anything either?

AntonLydike commented:
I'm going to move this discussion to someplace that is better suited for this tomorrow!

And I must admit, I don't really have any completely sound reasoning for those tests either. I mainly did this because I was lazy and did not understand the utility that more "fine grained" tests would provide.

I think my idea was to bundle a bunch of "trivial" tests into one bigger test (that also just happened to be the thing I used during the development of the lowering) to make it easier to write the test.

This results in some nice aspects:

 - If the the syntax of an operation is changed (and thereby the `get` method), we immediately have a test showing a use case for this op with context
 - It serves as a showcase for this dialect
 - It's an "easy" way to get a lot of  coverage

But it's also kind of flawed:

 - Using tests to show how to use a dialect is imo suboptimal, because ideally you'd do that in a separate place (notebook?) with a lot more explanations and context, and maybe even a complete story. (this is pretty debatable though, the more I think about it the more I like the concept of having dialect usecase examples be the tests, because that's how we write all our other tests as well, right?)
 - Constructing a big module, verifying it, printing it, and then comparing that output to a pre-computed value, just to test the operation of some `get` and `from_...` methods feels very convoluted to me.

georgebisbas commented:
Of course, we are not using tests to educate on dialects.
But the amount of resources for writing notebooks (work hours of the team needed to do this) is a lot.
Ideally, we should have a notebook for every dialect that shows some simple examples.
On the other side, testing bits of the dialect in a reasonable manner of operation-combos that make sense is not bad!

AntonLydike commented:
@georgebisbas I completely agree with that, I think having tests that cover actual usecases is good for a multitude of reasons. I moved this discussion over to https://github.com/xdslproject/xdsl/discussions/471
